src/Rings/mpoly-graded.jl

```diff
@@ -1612,6 +1612,386 @@ function Base.show(io::IO, h::HilbertData)
   print(io, "Hilbert Series for $(h.I), data: $(h.data), weights: $(h.weights)")  ###new
 end
 
+########################################################################
+# Compute the Hilbert series from the monomial leading ideal using 
+# different bisection strategies along the lines of 
+# [Kreuzer, Robbiano: Computational Commutative Algebra 2, Springer]
+# Section 5.3.
+########################################################################
+exp_vec(f::MPolyElem) = first(exponent_vectors(f))
+
+_divides(a::Vector{Int}, b::Vector{Int}) = all(k->(a[k]>=b[k]), 1:length(a))
+
+function _gcd(a::Vector{Vector{Int}}) 
+  length(a) == 1 && return a[1]
+  n = length(a[1]) # assumed to be the length of all entries of a
+  return [minimum([a[i][j] for i in 1:length(a)]) for j in 1:n]
+end
+
+function _are_pairwise_coprime(a::Vector{Vector{Int}})
+  length(a) <= 1 && return true
+  n = length(a)
+  for i in 1:n-1
+    for j in i+1:n
+      all(x->(x==0), _gcd([a[i], a[j]])) || return false
+    end
+  end
+  return true
+end
+
+### Assume that a is minimal. We divide by `pivot` and return 
+# a minimal set of generators for the resulting monomial ideal.
+function _divide_by(a::Vector{Vector{Int}}, pivot::Vector{Int};
+    # short exponent vectors where the k-th bit indicates that the k-th 
+    # exponent is non-zero.
+    shorts = [sum([(e[k]>0 ? 1 : 0) << (k-1) for k in 1:length(e)]) for e in a]
+  )
+  # The good ones will contribute to a minimal generating 
+  # set of the lhs ideal.
+  #
+  # The bad monomials come from those which hop over the boundaries of 
+  # the monomial diagram by the shift. Their span has a new 
+  # generator which is collected in `bad` a priori. It is checked 
+  # whether they become superfluous and if not, they are added to 
+  # the good ones.
+  good = similar(a, 0)
+  good_shorts = similar(shorts, 0)
+  bad = similar(a, 0)
+  bad_shorts = similar(shorts, 0)
+  short_piv = sum([(pivot[k]>0 ? 1 : 0) << (k-1) for k in 1:length(pivot)])
+  for (e, s) in zip(a, shorts)
+    next = e-pivot
+    # The first condition checks whether short_piv has a non-zero entry 
+    # where s does not. The inner expression evaluates to zero in the 
+    # affirmative case and pivot can not divide e.
+    if iszero(short_piv & ~ s) && all(k->(e[k]>=pivot[k]), 1:length(e)) # tuned version of divides(e, pivot)
+      push!(good, next)
+      push!(good_shorts, sum([(next[k]>0 ? 1 : 0) << (k-1) for k in 1:length(next)]))
```

Comment:
Did you try first applying my hints then measure, before adding these "shorts"? Because that code is again extremely inefficient and in this form likely constitutes a slowdown, not a speed up...


(The whole idea was to use bitmasks in an Int to be super fast; adding more vectors like this is several orders of magnitude slower. I would suggest to remoece the "shorts" again completely for now, and later Tommy or me can perhaps help adding them back in the intended form)

Reply:
To explain what is bad about this, observe that *must* allocate a `Vector{Int}`, which then is summed up. This takes up a ton of memory. Optimal code would "allocate" a single `Int` and then directly set bits in it.

Alas, let's not get hold up on this, let's just remove it for now. I'll write more in a longer comment in a moment